### PR TITLE
Reactivate Travis-CI

### PR DESCRIPTION
**Pull request by ocefpaf**
This is a substitute for #74 
 
I made several small commits to make it easier to review.
It should be easier for the reviewer to navigate the commits in order.

Note that the tests are now running, but not passing.
Most of them no longer make sense b/c they are outdated and disconnected from the latest code.

We should not only write new meaningful tests, but also write them together with the new code.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,67 +1,71 @@
-git blanguage: python
```
Comment (ocefpaf) [inline .travis.yml]: Probably a typo that prevented Travis-CI from running all this time.

File: .travis.yml [hunk 2]
```diff
@@ -1,67 +1,71 @@
...
-sudo: required
-# if the https://travis-ci.org/ODM2/ODM2PythonAPI/requests ever says: missing config
```
Comment (ocefpaf) [inline .travis.yml]: We don't really need `sudo` and without it Travis-CI can run in a faster using the container infrastructure. 

PS: I removed all the unnecessary comments.

File: .travis.yml [hunk 3]
```diff
@@ -1,67 +1,71 @@
...
-# validate at: http://lint.travis-ci.org/
-python:
```
Comment (ocefpaf) [inline .travis.yml]: Changed this to a test matrix definition with only Python 2.7 and 3.6.
Python 3.6 is unsupported now but I aim to change that soon.

File: .travis.yml [hunk 4]
```diff
@@ -1,67 +1,71 @@
...
-#  - "2.6"
-  - "2.7"
-#  - "3.2"
-#  - "3.3"
-#  - "3.4"
-#  - "3.5"
-#  - "3.5-dev" # 3.5 development branch
-#  - "nightly" # currently points to 3.6-dev
-# command to install dependencies
-cache:
```
Comment (ocefpaf) [inline .travis.yml]: To avoid a big PR I will add a caching mechanism in another PR as soon as this is merged.

File: .travis.yml [hunk 5]
```diff
@@ -1,67 +1,71 @@
...
-      - libc6
       - e2fsprogs
+      # FIXME: the dependencies below can probably be replaced by conda.
+      - libc6
+      - freetds-dev
+      - freetds-bin
+      - unixodbc
+      - unixodbc-dev
       - mysql-client
+      # FIXME: I don't see the need for these anywhere.
```
Comment (ocefpaf) [inline .travis.yml]: I don't really think we need any of these packages from the package manager.
Can someone with a more intimate knowledge of `odm2api` comment on this?

File: .travis.yml [hunk 6]
```diff
@@ -1,67 +1,71 @@
...
   - ./ci-helpers/mysql_setup.sh
   - ./ci-helpers/postgres_setup.sh
   - ./ci-helpers/travis/freetds.sh
 
-
 before_install:
-# python -m pip makes the install go into the virtualenv
-  - sudo pip install --upgrade pip
-  - sudo python -m pip install pandas
-  - export PYMSSQL_BUILD_WITH_BUNDLED_FREETDS=1;sudo python -m  pip install pymssql
```
Comment (ocefpaf) [inline .travis.yml]: I removed this complex pip installation to a more straight forward and reproducible `conda` env.
However, we may be losing some of the functionalities here. Unfortunately I cannot assess that without more comprehensive test.

For example, why do we need `PYMSSQL_BUILD_WITH_BUNDLED_FREETDS`?

File: .travis.yml [hunk 7]
```diff
@@ -1,67 +1,71 @@
...
+  - conda update --yes --all
+  - conda config --add channels odm2 --force
+  - conda create --yes -n TEST python=$TRAVIS_PYTHON_VERSION --file requirements.txt --file requirements-dev.txt
+  - source activate TEST
 
-# don't forget to open up the azure mssql server to these addreses
-# https://docs.travis-ci.com/user/ip-addresses/
+# Test source distribution.
+install:
+  - python setup.py sdist && version=$(python setup.py --version) && pushd dist  && pip install odm2api-${version}.tar.gz && popd
```
Comment (ocefpaf) [inline .travis.yml]: This is important to know if the source tarball is installable. We should not test against the source code tree.

File: .travis.yml [hunk 8]
```diff
@@ -1,67 +1,71 @@
...
 
-# command to run tests
 script:
+  - if [[ $TEST_TARGET == 'default' ]]; then
+      cp -r tests /tmp && cd /tmp ;
+      py.test -vv tests ;
```
Comment (ocefpaf) [inline .travis.yml]: This ensure that we are running the tests from the installed version and not the source code tree.

File: .travis.yml [hunk 9]
```diff
@@ -1,67 +1,71 @@
...
+    fi
 
-  - py.test
-
+  - if [[ $TEST_TARGET == 'coding_standards' ]]; then
+      find . -type f -name "*.py" ! -name 'conf.py' | xargs flake8 --max-line-length=100 ;
```
Comment (ocefpaf) [inline .travis.yml]: The coding standards tests, together with the Python 3 tests, are in the `allow_failures` section at the moment.
I plan to send small PRs to fix that and make these tests first class citizens here.

File: requirements-dev.txt [hunk 10]
```diff
@@ -0,0 +1,14 @@
+pytest
+pytest-cov
+mock
+pymysql
+pymssql
+psycopg2
+pymssql
+psycopg2
+flake8
+# FIXME: I am not sure these are needed.
```
Comment (ocefpaf) [inline requirements-dev.txt]: This needs a careful review.
I see packages dependencies all over the place and they don't really seems to be used.
I do know that, like for the `wofpy` case, people expected these to be installed even when the module in question is not using them directly.

If that is the case we should create an integration test env to add and test those and the expected functionality.

File: requirements_tests.txt [hunk 11]
```diff
@@ -1,15 +0,0 @@
--r requirements.txt
```
Comment (ocefpaf) [inline requirements_tests.txt]: Renamed to a more common name and removed the pip-ism syntax to make it package manager agnostic.

File: setup.py [hunk 12]
```diff
@@ -5,133 +5,51 @@
...
-
-    # To provide executable scripts, use entry points in preference to the
-    # "scripts" keyword. Entry points provide cross-platform support and allow
-    # pip to create the appropriate form of executable for the target platform.
-
-    # entry_points={
-    #     'console_scripts': [
-    #         'sample=sample:main',
-    #     ],
-    # },
```
Comment (ocefpaf) [inline setup.py]: Mostly cleanups. The comments are not really helpful and get in the way of reading the code here.

**Comment by emiliom** [comment]
@sreeder and @lsetiawan, can you review this PR from @ocefpaf? I suspect some areas may be outside your individual expertise, but we can work together on this with @ocefpaf's guidance. I won't be able to provide input until tomorrow (Thursday), and then @ocefpaf will be traveling starting on Friday.

Thanks!

File: setup.py [hunk 13]
```diff
@@ -5,133 +5,51 @@
...
 setup(
     name='odm2api',
-
-    # Versions should comply with PEP440.  For a discussion on single-sourcing
-    # the version across setup.py and the project code, see
-    # https://packaging.python.org/en/latest/single_source_version.html
     version=versioneer.get_version(),
-
-    description='A Python-based application programmers interface for the Observations Data converter 2 (ODM2) ',
+    description='Python interface for the Observations Data converter 2 (ODM2)',
```
Comment (lsetiawan) [inline setup.py]: Isn't ODM2 stands for Observations Data Model 2?

File: setup.py [hunk 14]
```diff
@@ -5,133 +5,51 @@
...
         'Operating System :: OS Independent',
-        # Specify the Python versions you support here. In particular, ensure
-        # that you indicate whether you support Python 2, Python 3 or both.
         'Programming Language :: Python :: 2.7',
         'Topic :: Software Development :: Libraries :: Python Modules',
         'Topic :: Scientific/Engineering'
     ],
-
-    # What does your project relate to?
     keywords='Observations Data converter ODM2',
```
Comment (lsetiawan) [inline setup.py]: Again, Observations Data Model 2 instead?

File: setup.py (hunk 13 above)
Comment (ocefpaf) [inline setup.py]: I only reduced the description that was already there, but I guess that we should also fix it to `Model`, right @emiliom?

File: .travis.yml (hunk 6 above)
Comment (sreeder) [inline .travis.yml]: Will the current setup test against an MSSQL database? I realize it is a Linux system so I am wondering if this was included to allow testing from MSSQL. If we aren't supporting that from here (which I am assuming we aren't) then we do not need it.

File: requirements-dev.txt (hunk 10 above)
Comment (sreeder) [inline requirements-dev.txt]: It looks like you have left in all of the dependencies that I usually install when running the API, so this looks good to me.

File: .travis.yml (hunk 6 above)
Comment (ocefpaf) [inline .travis.yml]: > Will the current setup test against an MSSQL database?

Probably not. I don't really know the unit/integration tests details and the current failures are happening due to other causes beyond the available services. (Also, databases are not my strong suit so I wouldn't know :grimacing:)

>  If we aren't supporting that from here (which I am assuming we aren't) then we do not need it.

@lsetiawan can you take a look at the tests and confirm that they don't requite a database service?

PS: note that we can do integration tests that relies on databases services in the future, but I would like to get a small set of tests working first.

Comment (lsetiawan) [inline .travis.yml]: > @lsetiawan can you take a look at the tests and confirm that they don't requite a database service?

Seems like odm2api tests are only using sqlite for the tests.

File: .travis.yml (hunk 5 above)
Comment (emiliom) [inline .travis.yml]: Which packages? If it's libproj-dev, libgeos-dev and libspatialiate-dev, I agree, they're not needed. I can't comment on `cmake`.

File: .travis.yml (hunk 6 above)
Comment (emiliom) [inline .travis.yml]: > PS: note that we can do integration tests that relies on databases services in the future, but I would like to get a small set of basic unittests working first.

I agree with that approach. We start with SQLite. 

Once we've built up to including RDBMS systems in the test suite, we can go back to considering MSSQL / FreeTDS testing in AppVeyor.

File: requirements-dev.txt (hunk 10 above)
Comment (emiliom) [inline requirements-dev.txt]: :+1: 

Except, you have `pymssql` and `psycopg2` listed twice

**Comment by emiliom** [comment]
Everything looks good to me, except for the duplicated listing of `pymssql` and `psycopg2` in `requirements-dev.txt`

**Comment by ocefpaf** [comment]
> Everything looks good to me, except for the duplicated listing of `pymssql` and `psycopg2` in `requirements-dev.txt`

Probably leftovers from a bad rebase. Latest commit should fix it.

**Pull request merged**